### PR TITLE
Add null-safety

### LINKED ISSUE #1002: Support null safety
This is a request to add support for [null safety](https://dart.dev/null-safety) to `package:dio`. This is a very popular package, so would be great to have null safety enabled early. See the [migration guide](https://dart.dev/null-safety/migration-guide) for details about enabling null safety.

### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### PR DESCRIPTION
**Pull request by harsimranmaan**
### New Pull Request Checklist

- [ ] I have read the [Documentation](https://pub.dartlang.org/packages/dio)
- [x] I have searched for a similar pull request in the [project](https://github.com/flutterchina/dio/pulls) and found none
- [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
- [ ] I have added the required tests to prove the fix/feature I am adding
- [ ] I have updated the documentation (if necessary)
- [x] I have run the tests and they pass

This merge request fixes / refers to the following issues: ...

### Pull Request Description

- Add conservative null-safety. Few tests are still failing with a common cause but looking for early feedback to see if this is the right direction to go. 
- Fix common  lint issues

Closes #1002

### CODE REVIEW

**Comment by harsimranmaan** [comment]
The tests fail with
```
00:05 +10 -5: test/request_test.dart: #test requests #test generic parameters [E]                             
  DioError [DioErrorType.DEFAULT]: type '() => Null' is not a subtype of type '(() => FutureOr<Socket>)?' of 'onTimeout'
  #0      Future.timeout (dart:async/future_impl.dart)
  #1      _ConnectionTarget.connect.<anonymous closure> (dart:_http/http_impl.dart:2222:37)
  #2      StackZoneSpecification._registerUnaryCallback.<anonymous closure>.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:126:26)
  #3      StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:208:15)
  #4      StackZoneSpecification._registerUnaryCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:126:14)
  #5      _rootRunUnary (dart:async/zone.dart:1194:47)
  #6      _CustomZone.runUnary (dart:async/zone.dart:1097:19)
  #7      _FutureListener.handleValue (dart:async/future_impl.dart:150:18)
  #8      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:703:45)
  #9      Future._propagateToListeners (dart:async/future_impl.dart:732:32)
  #10     Future._completeWithValue (dart:async/future_impl.dart:536:5)
  #11     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:574:7)
  #12     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:208:15)
  #13     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:116:48)
  #14     _rootRun (dart:async/zone.dart:1186:13)
  #15     _CustomZone.run (dart:async/zone.dart:1090:19)
  #16     _CustomZone.runGuarded (dart:async/zone.dart:994:7)
  #17     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1034:23)
  #18     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
  #19     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
  #20     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:120:13)
  #21     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:185:5)
  
  package:dio/src/dio.dart 875:17   DioMixin._request._errorInterceptorWrapper.<fn>.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                        Future.then
  package:dio/src/dio.dart 873:18   DioMixin._request._errorInterceptorWrapper.<fn>.<fn>
  package:dio/src/dio.dart 1127:22  DioMixin.checkIfNeedEnqueue
  package:dio/src/dio.dart 870:16   DioMixin._request._errorInterceptorWrapper.<fn>
  dart:async                        _completeOnAsyncError
  package:dio/src/dio.dart          DioMixin._dispatchRequest
  ===== asynchronous gap ===========================
  dart:async                        Future.catchError
  package:dio/src/dio.dart 907:23   DioMixin._request.<fn>
  dart:collection                   ListMixin.forEach
  package:dio/src/dio.dart 906:18   DioMixin._request
  package:dio/src/dio.dart 764:12   DioMixin.request
  package:dio/src/dio.dart 368:12   DioMixin.get
  test/request_test.dart 136:28     main.<fn>.<fn>
  test/request_test.dart 132:38     main.<fn>.<fn>
  
  OS Error: Connection reset by peer, errno = 54
  dart:isolate  _RawReceivePortImpl._handleMessage
```
Haven't had a chance to dig deep for the underlying cause yet.

File: dio/lib/src/transformer.dart
Comment (kuhnroyal) [inline dio/lib/src/transformer.dart:154]: Is this conversion needed?

Comment (venkatd) [reply dio/lib/src/transformer.dart:154]: Don't think it should be necessary since we're doing a null check already, no?

Comment (harsimranmaan) [reply dio/lib/src/transformer.dart:154]: Yes. I think VS tooling hasn't caught up yet for proper null safety checks. At the time of writing, I couldn't get the IDE to not complain here

Comment (venkatd) [reply dio/lib/src/transformer.dart:154]: @harsimranmaan I think this a limitation of the language and nothing to do with VS tooling. You will have to declare a local variable like `final decodeCallback = jsonDecodeCallback` and use that.

See here for more information: https://stackoverflow.com/questions/65456958/dart-null-safety-doesnt-work-with-class-fields

**Comment by miDeb** [comment]
@harsimranmaan the test failure seems to be caused by the dart sdk itself, I've opened https://github.com/dart-lang/sdk/issues/44895.

File: dio/lib/src/transformer.dart
Comment (mit-mit) [reply dio/lib/src/transformer.dart:154]: Yes, we currently cannot promote on class members. Using a local variable is a good workaround.

**Comment by miDeb** [comment]
@harsimranmaan the fix in `dart:io` has landed in the flutter master channel, and the tests pass again (well, they don't all pass on my machine but I believe that's because I'm on linux, so please check). I'd also have a few suggestions, but I can also open a follow up PR once this is merged. @wendux could you take a look?

**Comment by harsimranmaan** [comment]
Tests pass now against the dart version with the fix for the timeout issue. There were line ending mismatch in `#test FormData`. Would be worth running the test suite on windows too.

File: dio/lib/src/dio.dart
Comment (miDeb) [inline dio/lib/src/dio.dart:944]: The check for the actual value of `options.receiveDataWhenStatusError` is missing
```suggestion
      if (statusOk || options.receiveDataWhenStatusError == true) {
```

**Comment by miDeb** [comment]
Now the test is failing in windows :laughing: (checked in a windows vm). I believe `.replaceAll('\r\n', '\n')` should not be called on Windows.

**Comment by harsimranmaan** [comment]
Fixed the tests. Should work on both windows and other OS now.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:85]: If you use an initializer list here
```dart
: queryParameters = queryParameters ?? {},
 super();
```

then queryParameters doesn't need to be `late`. The reason why we can't use a default parameter here is that default parameters need to be `const`, which would mean that you can't modify the map afterwards anymore.
(See [the section about constructors in the language tour](https://dart.dev/guides/language/language-tour#constructors) for more information about initializer lists.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:57]: It is easier here to just use a default parameter instead of doing the check inside of the constructor.
```suggestion
    this.baseUrl = '',
```
You can then also remove `late` from `baseUrl`.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:117]: You can revert the change on this line, using `??` is just syntactic sugar for what you wrote.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:116]: Using `Map.from` probably had the intention to clone the map. I guess you should add it back

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:193]: Same as above - `Map.from` is probably required.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:215]: Same as with the `BaseOptions` constructor. 
```suggestion
    this.path = '',
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:217]:
```suggestion
    this.baseUrl = '',
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:247]: Again, queryParameters can be in the initializer list and doesn't need to be late.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:288]: `Map.from` is probably required.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:351]: Here we can again use default parameters to simplify the code.
```suggestion
    this.receiveDataWhenStatusError = true,
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:352]:
```suggestion
    this.followRedirects = true,
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:353]:
```suggestion
    this.maxRedirects = 5,
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:348]:
```suggestion
    this.contentType,
```

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:358]: If you move `headers` and `extra` to the initializer list they don't have to be `late`.

File: dio/lib/src/options.dart
Comment (miDeb) [inline dio/lib/src/options.dart:417]: `receiveDataWhenStatusError`, `followRedirects` and `maxRedirects` don't need to be nullable.

File: dio/lib/src/dio.dart
Comment (miDeb) [inline dio/lib/src/dio.dart:49]: `options` can be non-nullable. I believe there are different ways to achieve that...
If you look around for its usages it is always assigned `BaseOptions()` as a default value, and every access uses `options!.foo`. I believe there is no reason to make it nullable.

File: dio/lib/src/dio.dart
Comment (miDeb) [inline dio/lib/src/dio.dart:57]: Bonus points if you can make `options`, `httpClientAdapter` and `transformer` not be `late`; It's just because they are always initialized right away in constructors, which in most cases can be somehow rewritten to not require `late`. I don't think that change is absolutely necessary, though.

**Review by miDeb — CHANGES_REQUESTED** [review]
Hope these suggestions make sense :)

File: dio/lib/src/options.dart
Comment (harsimranmaan) [reply dio/lib/src/options.dart:417]: This is a bit tricky as it affect merge-ability. Would not be possible to detect if the source had the bool/int value set or default value of true is being used.

File: dio/lib/src/options.dart
Comment (harsimranmaan) [reply dio/lib/src/options.dart:348]: contentType is not a class field bur set/get

**Comment by harsimranmaan** [comment]
Thanks for the suggestions @miDeb. I have addressed the feedback. I have started to like dart now :)

File: dio/lib/src/options.dart
Comment (miDeb) [reply dio/lib/src/options.dart:408]: I think you are referring to the `mergeOptions` or `merge` methods, right? The thing is, it was impossible to create a `_RequestConfig` with a null value for these fields. [Its constructor used to assign default values if you passed null](https://github.com/flutterchina/dio/pull/1026/commits/d3a208de4bf217663eaa3c83c33cf8d64daaaf56#diff-bed0ea39c9af0b71c65fb305e8c68236e0292cad32dd6693fe8092aefcdfb6a7L360). The way it is now, it would be possible to pass a null value, but my point is it was not possible before. It was only possible to re-set the default value and assign null after the object was created. I just felt that was unlikely to happen, but I obviously don't know how all the usage of `dio` looks like.
I do however think that the original author was confused about where a null value represents a default and where there are non-null default values. I noticed this in code I have written as well, null safety really helps here!

File: dio/lib/src/options.dart
Comment (miDeb) [reply dio/lib/src/options.dart:343]: You're obviously right!

**Review by miDeb — APPROVED** [review]
I'm obviously not a mantainer here, but these changes now look good to me! Thanks for all your effort @harsimranmaan!

**Comment by acoutts** [comment]
LGTM!

**Comment by acoutts** [comment]
Don't forget to bump the version to next major release as per the null safety migration recommendations.

**Comment by kuhnroyal** [comment]
Yea that needs to be done by the absent maintainer :)

**Comment by miDeb** [comment]
One more thing to keep in mind is whether the fix to https://github.com/dart-lang/sdk/issues/44895 has landed in all null-safe release channels.

**Comment by TitanKing** [comment]
Thanks for now I will use your package @harsimranmaan, appreciate your efforts. 

This might need to be created as a new package.

**Comment by aarontwf** [comment]
Does the cookie_manager plugin also need updating?

**Comment by miDeb** [comment]
> Does the cookie_manager plugin also need updating?
That requires this pr and [cookie_jar](https://github.com/flutterchina/cookie_jar/pull/24) first, but otherwise yes.

**Comment by masseelch** [comment]
For everyone landing here: If you want to use this PR until it gets merged add this to your `pubspec.yaml`:

```yaml
dependency_overrides:
  dio: # https://github.com/flutterchina/dio/issues/1002
    git:
      url: git@github.com:harsimranmaan/dio.git
      ref: null-safety
      path: dio
```

Remove it once it gets merged.

**Comment by j4qfrost** [comment]
@wendux Thanks for the merge on cookie_jar. Do you mind taking a look this too?

**Pull request merged**
